redoctober_test.go

```diff
@@ -0,0 +1,684 @@
+// these tests first require that you build this project and have a recent
+// binary of it in either the /redoctober folder or in $GOPATH/bin/redoctober.
+
+package main
+
+import (
+	"bytes"
+	"crypto/tls"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"os/exec"
+	"testing"
+	"time"
+
+	"github.com/cloudflare/redoctober/core"
+)
+
+const baseURL = "https://localhost:8080/"
+
+var (
+	// Inputs that are used in multiple test cases are put here to reduce verbosity.
+	createVaultInput = &core.CreateRequest{Name: "Alice", Password: "Lewis"}
+
+	createUserInput1 = &core.CreateUserRequest{Name: "Bill", Password: "Lizard"}
+	createUserInput2 = &core.CreateUserRequest{Name: "Cat", Password: "Cheshire"}
+	createUserInput3 = &core.CreateUserRequest{Name: "Dodo", Password: "Dodgson"}
+
+	delegateInput1 = &core.DelegateRequest{
+		Name: createUserInput1.Name,
+		Password: createUserInput1.Password,
+		Time: "2h34m",
+		Uses: 1,
+	}
+	delegateInput2 = &core.DelegateRequest{
+		Name: createUserInput2.Name,
+		Password: createUserInput2.Password,
+		Time: "2h34m",
+		Uses: 1,
+	}
+
+	encryptInput     = &core.EncryptRequest{
+		Name:     createVaultInput.Name,
+		Password: createVaultInput.Password,
+		Owners:   []string{createUserInput1.Name, createUserInput2.Name},
+		Data:     []byte("V2h5IGlzIGEgcmF2ZW4gbGlrZSBhIHdyaXRpbmcgZGVzaz8K"),
+	}
+)
+
+func init() {
+	// Because the certificate being used is self-signed, InsecureSkipVerify must be enabled
+	// to avoid the POST requests from failing.
+	cfg := &tls.Config{
+		InsecureSkipVerify: true,
+	}
+
+	http.DefaultClient.Transport = &http.Transport{
+		TLSClientConfig: cfg,
+	}
+}
+
+func setup(t *testing.T) (cmd *exec.Cmd) {
+	// Look for the redoctober binary in current directory and then in $GOPATH/bin
+	binaryPath, err := exec.LookPath("./redoctober")
+	if err != nil {
+		goPathBinary := fmt.Sprintf("%s/bin/redoctober", os.Getenv("GOPATH"))
+		binaryPath, err = exec.LookPath(goPathBinary)
+		if err != nil {
+			t.Fatalf(`Could not find redoctober binary at "./redoctober" or "%s"`, goPathBinary)
+		}
+	}
+
+	cmd = exec.Command(binaryPath, "-addr=localhost:8080", "-certs=testdata/server.crt",
+		"-keys=testdata/server.pem", "-vaultpath=memory")
+
+	if err := cmd.Start(); err != nil {
+		t.Fatalf("Error running redoctober command, %v", err)
+	}
+
+	// Give the server time to start up.
+	time.Sleep(500 * time.Millisecond)
+
+	return
+}
+
+func teardown(t *testing.T, cmd *exec.Cmd) {
+	err := cmd.Process.Kill()
+	if err != nil {
+		t.Fatalf("Error killing the redoctober server, %v", err)
+	}
+}
+
+func post(api string, v interface{}) (respBytes []byte, response *http.Response, err error) {
+	jsonBytes, err := json.Marshal(v)
+	if err != nil {
+		return
+	}
+
+	buffer := bytes.NewBuffer(jsonBytes)
+	response, err = http.Post(baseURL+api, "text/json", buffer)
+	if err != nil {
+		return
+	}
+
+	respBytes, err = ioutil.ReadAll(response.Body)
```

Comment:
You should close the response body after reading it; you can do this before checking the error; alternatively, defer the response body close after posting.